library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,908 @@
+            number: patch_number,
+        }
+    }
+
+    /// Checks that the patch with the given number:
+    ///   - Has an artifact on disk
+    ///   - That artifact on disk is the same size it was when it was installed
+    ///
+    /// Returns Ok if the patch is bootable, or an error if it is not.
+    fn validate_patch_is_bootable(&self, patch: &PatchMetadata) -> Result<()> {
+        let artifact_path = self.patch_artifact_path(patch.number);
+        if !Path::exists(&artifact_path) {
+            bail!(
+                "Patch {} does not exist at {}",
+                patch.number,
+                artifact_path.display()
+            );
+        }
+
+        let artifact_size_on_disk = std::fs::metadata(&artifact_path)?.len();
+        if artifact_size_on_disk != patch.size {
+            bail!(
+                "Patch {} has size {} on disk, but expected size {}",
+                patch.number,
+                artifact_size_on_disk,
+                patch.size
+            );
+        }
+
+        Ok(())
+    }
+
+    fn delete_patch_artifacts(&mut self, patch_number: usize) -> Result<()> {
+        info!("Deleting patch artifacts for patch {}", patch_number);
+
+        let patch_dir = self.patch_dir(patch_number);
+
+        std::fs::remove_dir_all(&patch_dir)
+            .map_err(|e| {
+                error!("Failed to delete patch dir {}: {}", patch_dir.display(), e);
+                e
+            })
+            .with_context(|| format!("Failed to delete patch dir {}", &patch_dir.display()))
+    }
+
+    /// Attempts to use the last successfully booted patch as the next boot patch. If the last successfully
+    /// booted patch is not bootable or has the same number as the patch we're falling back from, we clear it.
+    fn try_fall_back_to_last_booted_patch(&mut self) {
+        if let Some(next_boot_patch) = self.patches_state.next_boot_patch {
+            // If we have a next_boot_patch that we're falling back from, delete its artifacts.
+            let _ = self.delete_patch_artifacts(next_boot_patch.number);
+            self.patches_state.next_boot_patch = None;
+
+            if let Some(last_boot_patch) = self.patches_state.last_booted_patch {
+                if last_boot_patch.number == next_boot_patch.number {
+                    // If the last booted patch is the same as the next boot patch, clear it.
+                    self.patches_state.last_booted_patch = None;
+                }
+            }
+        }
+
+        if let Some(last_boot_patch) = self.patches_state.last_booted_patch {
+            if self.validate_patch_is_bootable(&last_boot_patch).is_ok() {
+                // If we think we can still boot from the last booted patch, set it as the next_boot_patch.
+                self.patches_state.next_boot_patch = Some(last_boot_patch);
+            } else {
+                self.patches_state.last_booted_patch = None;
+                let _ = self.delete_patch_artifacts(last_boot_patch.number);
```

Comment:
I presume this validation of the next_boot_patch happens elsewhere?  Is this just a copy of some of the install logic?